sql/catalog/descs,multiregionccl: allow system database to be made MR

ajwerner opened the pull request:
This change is needed in order to convert system database tables to
regional by row.

This PR removes some logic which prevented the system database from being
properly modified. It also adds a test which ensures that the changes happen
as expected.

Epic: CRDB-18596

Release note: None

ajwerner committed:
descs: special case the system public schema when hydrating mutable types

This change is needed in order to convert system database tables to
regional by row.

Release note: None

ajwerner committed:
sql/catalog,multiregionccl: ensure system database can be modified

This PR removes some logic which prevented the system database from being
properly modified. It also adds a test which ensures that the changes happen
as expected.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/91970)
<!-- Reviewable:end -->

ajwerner commented:
@JeffSwenson you're going to need this

ajwerner committed:
dbdesc: set the system database descriptor version to 1 if 0

There are very old versions out there which do not have versions in their
descriptors. Long ago when these versions were introduced, we incremented
the version on change. The only descriptor we never actually wrote to was
the system descriptor. Those days are over. To account for that, we set
its version to 1 implicitly during post-deserialization. This fixes the
version upgrade tests.

Release note: None

JeffSwenson commented:
Thanks! I'm going to cherry pick this into my sql_instances change so that I can send it out for review.

postamar commented on pkg/sql/catalog/descs/hydrate.go:
```diff
@@ -120,6 +121,19 @@ func makeMutableTypeLookupFunc(
 	}
 	mutableLookupFunc := func(ctx context.Context, id descpb.ID) (catalog.Descriptor, error) {
+		// This special case exists to deal with the desire to use enums in the
+		// system, and the fact that the hydration contract is such that when we
```
nit: did you mean "in the system database"?

postamar commented on pkg/sql/catalog/descs/hydrate.go:
```diff
@@ -120,6 +121,19 @@ func makeMutableTypeLookupFunc(
+				Required:    true,
+				AvoidLeased: true,
+			})
+		}
```
I worry that, whenever we do endow system.public with an actual descriptor, removing this code will be overlooked.  I'm not sure what to do about this, though. Perhaps me referencing https://github.com/cockroachdb/cockroach/issues/89849 is enough.

postamar commented on pkg/sql/catalog/dbdesc/database_desc_builder.go:
```diff
@@ -107,6 +107,16 @@ func (ddb *databaseDescriptorBuilder) RunPostDeserializationChanges() (err error
 
+	// This should only every happen to the system database. Unlike many other
+	// post-deserialization changes, this does not need to last forever to
+	// support restores. It can be removed in 23.2.
```
> It can be removed in 23.2.

Why is that the case?

postamar commented on pkg/ccl/multiregionccl/regional_by_row_system_database_test.go:
```diff
@@ -0,0 +1,79 @@
+
+// This is a rather contrived test to demonstrate that one can successfully,
+// albeit with some arm-twisting, convince the sql layer to make regional-
+// by-row tables in the system database.
```
Shouldn't this PR include a new CCL logic test config in which the system database is multi-region? Seems like it'd be worth doing because it'd be nice to add more test cases, then again I have no idea what kind of effort it involves. You be the judge.

postamar reviewed (approved):
LGTM. I _think_ you took care of all of the gotchas that made the system database special. At least, I can't think of any more.

ajwerner reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91970#-:-NH6fbyhBJJZ_JPo6aQf:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @postamar)

___
*[`pkg/sql/catalog/dbdesc/database_desc_builder.go` line 112 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91970#-NGyc8aB9FZ_epJRCFFr-r3-112:-NH6aL7B7NCEw4v6pMLp:bjtxueh) ([raw file](https://github.com/cockroachdb/cockroach/blob/76b2475d12bf8febb7e760b276dab16373fc8e23/pkg/sql/catalog/dbdesc/database_desc_builder.go#L112)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

> It can be removed in 23.2.

Why is that the case?
</blockquote></details>

When I typed this, my thinking was that we'd have a migration in 23.1 which upgrades the system database. Nothing right now ensures that. I've reworded this comment. 
___
*[`pkg/sql/catalog/descs/hydrate.go` line 125 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91970#-NGyMwk0EhLSbYQr67w4-r3-125:-NH6apEP1Y5UgWe6xZRa:b60r506) ([raw file](https://github.com/cockroachdb/cockroach/blob/76b2475d12bf8febb7e760b276dab16373fc8e23/pkg/sql/catalog/descs/hydrate.go#L125)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

nit: did you mean "in the system database"?
</blockquote></details>

Yes, done.
___
*[`pkg/sql/catalog/descs/hydrate.go` line 136 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91970#-NGyMwk0EhLSbYQr67w4-r3-136:-NH6azAT75dplJZ_VYyN:b-c65e1w) ([raw file](https://github.com/cockroachdb/cockroach/blob/76b2475d12bf8febb7e760b276dab16373fc8e23/pkg/sql/catalog/descs/hydrate.go#L136)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I worry that, whenever we do endow system.public with an actual descriptor, removing this code will be overlooked.  I'm not sure what to do about this, though. Perhaps me referencing https://github.com/cockroachdb/cockroach/issues/89849 is enough.
</blockquote></details>

Probably it is good enough. The good news is it's still legit in that because the only thing we would need to see out of a mutable descriptor for this descriptor might be a changed name, and I don't think we'll permit changing the name of the public schema (we already don't).
___
*[`pkg/ccl/multiregionccl/regional_by_row_system_database_test.go` line 24 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91970#-NGyMwk11c4y1mwKME1a-r3-24:-NH6bD3u8s_fg9qCAcrl:bqzi7b7) ([raw file](https://github.com/cockroachdb/cockroach/blob/76b2475d12bf8febb7e760b276dab16373fc8e23/pkg/ccl/multiregionccl/regional_by_row_system_database_test.go#L24)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Shouldn't this PR include a new CCL logic test config in which the system database is multi-region? Seems like it'd be worth doing because it'd be nice to add more test cases, then again I have no idea what kind of effort it involves. You be the judge.
</blockquote></details>

I think such a thing will be good. I'll add it in a later PR. Filed https://github.com/cockroachdb/cockroach/issues/92090. 


<!-- Sent from Reviewable.io -->

ajwerner commented:
bors r+
